framework/OpenMod.Core/Localization/SmartFormatOptions.cs

```diff
@@ -0,0 +1,131 @@
+
+        // copy formatters and sources
+        m_Formatters = defaultFormatter.GetFormatterExtensions().ToList();
+        m_Sources = defaultFormatter.GetSourceExtensions().ToList();
+
+        // Smartformatter will be used in threading, so make sure it is thread safe mode is set
+        // more info: https://github.com/axuno/SmartFormat/wiki/Async-and-Thread-Safety
+        SmartSettings.IsThreadSafeMode = true;
+
+        // adds support for custom extensions, but removes the string.Format compatibility
+        m_SmartSettings.StringFormatCompatibility = false;
+    }
+
+    /// <summary>
+    /// Adds <see cref="IFormatter"/> extension to the formatters list
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="formatter"><see cref="IFormatter"/> instance</param>
+    /// <returns><see langword="true"/>, if the extension was added.</returns>
+    public bool TryAddFormatter<T>(T formatter) where T : class, IFormatter
+    {
+        CheckInitialized();
+
+        if (m_Formatters.OfType<T>().FirstOrDefault() is not null)
+        {
+            return false;
+        }
+
+        m_Formatters.Add(formatter);
+        return true;
+    }
+
+    /// <summary>
+    /// Removes <see cref="IFormatter"/> extension from the formatters list
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <returns><see langword="true"/>, if the extension was removed.</returns>
+    public bool TryRemoveFormatter<T>() where T : class, IFormatter
+    {
+        CheckInitialized();
+
+        return m_Formatters.RemoveAll(f => f is T) > 0;
+    }
+
+    /// <summary>
+    /// Adds <see cref="ISource"/> extension to the source list
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="source"><see cref="ISource"/> instance</param>
+    /// <returns><see langword="true"/>, if the extension was added.</returns>
+    public bool TryAddSource<T>(T source) where T : class, ISource
+    {
+        CheckInitialized();
+
+        if (m_Sources.OfType<T>().FirstOrDefault() is not null)
+        {
+            return false;
+        }
+
+        m_Sources.Add(source);
+        return true;
+    }
+
+    /// <summary>
+    /// Removes <see cref="ISource"/> extension from the source list
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <returns><see langword="true"/>, if the extension was removed.</returns>
+    public bool TryRemoveSource<T>() where T : class, ISource
+    {
+        CheckInitialized();
+
+        return m_Sources.RemoveAll(f => f is T) > 0;
+    }
+
+    public void SetLocalizationProvider(ILocalizationProvider? localizationProvider)
+    {
+        CheckInitialized();
+
+        m_SmartSettings.Localization.LocalizationProvider = localizationProvider;
+    }
+
+    private void CheckInitialized()
+    {
+        if (m_IsInitialized)
+        {
+            throw new InvalidOperationException("SmartFormatter already initialized");
+        }
```

Comment:
I think this should be inverted:
```cs
if (!m_IsInitialized) 
{
   throw...
}
```

Reply:
No, it should be like this to prevent using `TryAdd...` if `GetSmartFormatter()` was used, because `SmartFormatter` already cached and cannot be modified.